### PR TITLE
Heap profiling

### LINKED ISSUE #15990: Programmatic heap profiling
## Reason
We support generating CPU profiles for every process through the environment variable `MOOSE_PROFILE_BASE`. We should allow something similar for heap profiling

## Design
I was hoping I could just duplicate what's done for CPU profiling by adding appropriate calls to `HeapProfilerStart` and `HeaProfilerStop` but while I'm told at the top of a simulation that heap profiling has started, no files are ever written.

## Impact
Allow easy heap profiling like we have for CPU profiling (which is awesome).

### PR DESCRIPTION
**Pull request by fdkong**
Closes #15990

### CODE REVIEW

File: modules/doc/content/application_development/profiling.md [hunk 1]
```diff
@@ -47,6 +47,22 @@ This will use the filename base you pass and append a suffix of the form
 rank/process. The above example would generate files `run1_0.prof`,
 `run1_1.prof`, `run1_2.prof`, ..., `run1_31.prof`.
 
+### Heap Profiling
+
+Similarly, you can do a heap profiling like this:
+
+```
+HEAP_PROFILE_INUSE_INTERVAL=104857600 MOOSE_HEAP_BASE=run1_ mpiexec -n 32 ./your-app_oprof -i input_file.i
```
Comment (lindsayad) [inline modules/doc/content/application_development/profiling.md:55]: Isn't this the default value? Does anything happen if you don't specify this environment value? I remember when I was trying to do this, I would not get any memory profiles dumped ever, but I wasn't specifying any environment variable values like `HEAP_PROFILE_INUSE_INTERVAL`

Comment (fdkong) [reply modules/doc/content/application_development/profiling.md:55]: The default is nothing. Right, you do not get anything if you do not specify `HEAP_PROFILE_INUSE_INTERVAL` or other variables

**Comment by fdkong** [comment]
Do not merge it now. I am going to try it out

File: framework/src/base/MooseApp.C [hunk 2]
```diff
@@ -343,13 +344,29 @@ MooseApp::MooseApp(InputParameters parameters)
...
+
+    _cpu_profiling = true;
     ProfilerStart(profile_file.c_str());
   }
+
+  // For Heap profiling, users need to have 'MOOSE_HEAP_BASE'
+  if (std::getenv("MOOSE_HEAP_BASE"))
+  {
+    static std::string profile_file =
+        std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank()) + ".prof";
```
Comment (lindsayad) [inline framework/src/base/MooseApp.C:364]: I don't think we want the `.prof` string in here do we? At least on my machine, libtcmalloc actually appends a `.heap` suffix on its own

**Review by lindsayad — DISMISSED** [review]

File: framework/src/base/MooseApp.C (hunk 2 above)
Comment (fdkong) [reply framework/src/base/MooseApp.C:364]: Ok, I will remove that

**Comment by lindsayad** [comment]
This is working really well for me after the `.mk` fixes. I love this. Thanks @fdkong

**Comment by fdkong** [comment]
OK, it works for me on HPC.

File: framework/src/base/MooseApp.C (hunk 2 above)
Comment (fdkong) [reply framework/src/base/MooseApp.C]: Done

**Review by lindsayad — COMMENTED** [review]
I have a commit on this now, so maybe we should have another reviewer

**Comment by fdkong** [comment]
@rwcarlsen Could you take a look? Thanks

**Review by rwcarlsen — APPROVED** [review]

**Comment by lindsayad** [comment]
Something I found today is that if tcmalloc symbols are ahead of libc symbols, then you get a bazillion valgrind errors when running valgrind (invalid frees, etc.). So that is unfortunate

File: modules/doc/content/application_development/profiling.md [hunk 3]
```diff
@@ -47,6 +47,22 @@ This will use the filename base you pass and append a suffix of the form
...
+```
```
Comment (YaqiWang) [inline modules/doc/content/application_development/profiling.md:56]: Maybe we are missing an empty line bellow? Check this out; https://mooseframework.inl.gov/docs/PRs/16224/site/application_development/profiling.html.

Comment (fdkong) [reply modules/doc/content/application_development/profiling.md:56]: Thanks. Good catch. I will fix that

**Pull request merged**